src/avro_copy.cpp

```diff
@@ -0,0 +1,498 @@
+	}
+}
+
+static void WriteAvroSink(ExecutionContext &context, FunctionData &bind_data_p, GlobalFunctionData &gstate_p,
+                          LocalFunctionData &lstate_p, DataChunk &input) {
+	auto &global_state = gstate_p.Cast<WriteAvroGlobalState>();
+	auto &local_state = lstate_p.Cast<WriteAvroLocalState>();
+
+	auto &datum_buffer = global_state.datum_buffer;
+	idx_t count = input.size();
+	idx_t offset_in_datum_buffer = 0;
+	for (idx_t i = 0; i < count; i++) {
+
+		//! Populate our avro value, estimating the size of the value as we go
+		idx_t value_size = 0;
+		for (idx_t col_idx = 0; col_idx < input.ColumnCount(); col_idx++) {
+			auto val = input.GetValue(col_idx, i);
+
+			const char *unused_name;
+			avro_value_t column;
+			if (avro_value_get_by_index(&local_state.value, col_idx, &column, &unused_name)) {
+				throw InvalidInputException(avro_strerror());
+			}
+			value_size += PopulateValue(&column, val);
+		}
+
+		//! Prepare the datum buffer for this row
+		idx_t length = datum_buffer.GetCapacity() - offset_in_datum_buffer;
+		if (value_size > length) {
+			//! This value is too big to fit into the remaining portion of the buffer
+			idx_t new_capacity = datum_buffer.GetCapacity();
+			new_capacity += value_size;
+			datum_buffer.ResizeAndCopy(NextPowerOfTwo(new_capacity));
+		}
+		avro_writer_memory_set_dest_with_offset(global_state.datum_writer, (const char *)datum_buffer.GetData(),
+		                                        datum_buffer.GetCapacity(), offset_in_datum_buffer);
+
+		while (avro_file_writer_append_value(global_state.file_writer, &local_state.value)) {
+			auto current_capacity = datum_buffer.GetCapacity();
+			datum_buffer.ResizeAndCopy(NextPowerOfTwo(current_capacity * 2));
+			avro_writer_memory_set_dest_with_offset(global_state.datum_writer, (const char *)datum_buffer.GetData(),
+			                                        datum_buffer.GetCapacity(), offset_in_datum_buffer);
+		}
+
+		offset_in_datum_buffer = avro_writer_tell(global_state.datum_writer);
+		avro_value_reset(&local_state.value);
+	}
+
+	auto &buffer = global_state.memory_buffer;
+	auto expected_size = avro_writer_tell(global_state.datum_writer);
+	expected_size += 16 + 4;
+	if (expected_size > buffer.GetCapacity()) {
+		//! Resize the buffer in advance, to prevent any need for resizing below
+		buffer.Resize(NextPowerOfTwo(expected_size));
+		avro_writer_memory_set_dest(global_state.writer, (const char *)buffer.GetData(), buffer.GetCapacity());
+	}
+
+	//! Flush the contents to the buffer, if it fails, resize the buffer and try again
+	while (avro_file_writer_flush(global_state.file_writer)) {
+		auto current_capacity = buffer.GetCapacity();
+		buffer.Resize(NextPowerOfTwo(current_capacity * 2));
+		avro_writer_memory_set_dest(global_state.writer, (const char *)buffer.GetData(), buffer.GetCapacity());
+	}
+
+	auto written_bytes = avro_writer_tell(global_state.writer);
+	global_state.WriteData(buffer.GetData(), written_bytes);
```

Comment:
maybe write data can return a return code? 
Then if the write fails we can return a nice error message here? Seems like `handle->Write((void *)data, size);` in `WriteAvroGlobalState::WriteData` does this